bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogManagerForEntryLogPerLedger.java

```diff
@@ -0,0 +1,463 @@
+     * Cache.cleanUp() method of the returned cache will also perform
+     * maintenance, but calling it should not be necessary with a high
+     * throughput cache. Only caches built with removalListener,
+     * expireAfterWrite, expireAfterAccess, weakKeys, weakValues, or
+     * softValues perform periodic maintenance.
+     */
+    @VisibleForTesting
+    void doEntryLogMapCleanup() {
+        ledgerIdEntryLogMap.cleanUp();
+    }
+
+    @VisibleForTesting
+    ConcurrentMap<Long, EntryLogAndLockTuple> getCacheAsMap() {
+        return ledgerIdEntryLogMap.asMap();
+    }
+    /*
+     * Returns writable ledger dir with least number of current active
+     * entrylogs.
+     */
+    @Override
+    public File getDirForNextEntryLog(List<File> writableLedgerDirs) {
+        Map<File, MutableInt> writableLedgerDirFrequency = new HashMap<File, MutableInt>();
+        writableLedgerDirs.stream()
+                .forEach((ledgerDir) -> writableLedgerDirFrequency.put(ledgerDir, new MutableInt()));
+        for (BufferedLogChannel logChannel : replicaOfCurrentLogChannels.values()) {
+            File parentDirOfCurrentLogChannel = logChannel.getLogFile().getParentFile();
+            if (writableLedgerDirFrequency.containsKey(parentDirOfCurrentLogChannel)) {
+                writableLedgerDirFrequency.get(parentDirOfCurrentLogChannel).increment();
+            }
+        }
+        @SuppressWarnings("unchecked")
+        Optional<Entry<File, MutableInt>> ledgerDirWithLeastNumofCurrentLogs = writableLedgerDirFrequency.entrySet()
+                .stream().min(Map.Entry.comparingByValue());
+        return ledgerDirWithLeastNumofCurrentLogs.get().getKey();
+    }
+
+    @Override
+    public void close() throws IOException {
+        Set<BufferedLogChannel> copyOfCurrentLogs = getCopyOfCurrentLogs();
+        for (BufferedLogChannel currentLog : copyOfCurrentLogs) {
+            EntryLogger.closeFileChannel(currentLog);
+        }
+    }
+
+    @Override
+    public void forceClose() {
+        Set<BufferedLogChannel> copyOfCurrentLogs = getCopyOfCurrentLogs();
+        for (BufferedLogChannel currentLog : copyOfCurrentLogs) {
+            EntryLogger.forceCloseFileChannel(currentLog);
+        }
+    }
+
+    @Override
+    void flushCurrentLogs() throws IOException {
+        Set<BufferedLogChannel> copyOfCurrentLogs = getCopyOfCurrentLogs();
+        for (BufferedLogChannel logChannel : copyOfCurrentLogs) {
+            /**
+             * flushCurrentLogs method is called during checkpoint, so metadata
+             * of the file also should be force written.
+             */
+            flushLogChannel(logChannel, true);
+        }
+    }
+
+    @Override
+    public BufferedLogChannel createNewLogForCompaction() throws IOException {
+        throw new UnsupportedOperationException(
+                "When entryLogPerLedger is enabled, transactional compaction should have been disabled");
```

Comment:
is this checked in configuration as well? 
i'd prefer crash on start with message about wrong configuration to crash after few hrs of running.

Reply:
+1

Reply:
yes, it is validated in ServerConfiguration.validate method - https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java#L2268